File: src/main/java/me/oddlyoko/odoo/models/reference/OdooModelReferenceContributor.java
package me.oddlyoko.odoo.models.reference;

import com.intellij.patterns.PlatformPatterns;
import com.intellij.patterns.PsiElementPattern;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiReference;
import com.intellij.psi.PsiReferenceContributor;
import com.intellij.psi.PsiReferenceProvider;
import com.intellij.psi.PsiReferenceRegistrar;
import com.intellij.util.ProcessingContext;
import com.jetbrains.python.PyTokenTypes;
import com.jetbrains.python.psi.PyDictLiteralExpression;
import com.jetbrains.python.psi.PyListLiteralExpression;
import com.jetbrains.python.psi.PySetLiteralExpression;
import com.jetbrains.python.psi.PyStringLiteralExpression;
import com.jetbrains.python.psi.PyTargetExpression;
import org.jetbrains.annotations.NotNull;

public class OdooModelReferenceContributor extends PsiReferenceContributor {
    public static final PsiElementPattern.Capture<? extends PsiElement> NAME_PATTERN =
            PlatformPatterns.psiElement(PyStringLiteralExpression.class).afterSiblingSkipping(
                    PlatformPatterns.psiElement().withElementType(PyTokenTypes.EQ),
                    PlatformPatterns.psiElement(PyTargetExpression.class).withName("_name")
            );
    public static final PsiElementPattern.Capture<? extends PsiElement> INHERIT_PATTERN =
            PlatformPatterns.psiElement(PyStringLiteralExpression.class).afterSiblingSkipping(
                    PlatformPatterns.psiElement().withElementType(PyTokenTypes.EQ),
                    PlatformPatterns.psiElement(PyTargetExpression.class).withName("_inherit")
            );
    public static final PsiElementPattern.Capture<? extends PsiElement> INHERIT_DICT_PATTERN =
            PlatformPatterns.psiElement(PyStringLiteralExpression.class).withParent(
                    PlatformPatterns.psiElement(PyListLiteralExpression.class).afterSiblingSkipping(
                            PlatformPatterns.psiElement().withElementType(PyTokenTypes.EQ),
                            PlatformPatterns.psiElement(PyTargetExpression.class).withName("_inherit")
                    )
            );
    public static final PsiElementPattern.Capture<? extends PsiElement> INHERITS_PATTERN =
            PlatformPatterns.psiElement(PyStringLiteralExpression.class).withAncestor(2,
                    PlatformPatterns.or(
                            PlatformPatterns.psiElement(PyDictLiteralExpression.class).afterSiblingSkipping(
                                    PlatformPatterns.psiElement().withElementType(PyTokenTypes.EQ),
                                    PlatformPatterns.psiElement(PyTargetExpression.class).withName("_inherits")
                            ),
                            PlatformPatterns.psiElement(PySetLiteralExpression.class).afterSiblingSkipping(
                                    PlatformPatterns.psiElement().withElementType(PyTokenTypes.EQ),
                                    PlatformPatterns.psiElement(PyTargetExpression.class).withName("_inherits")
                            )
                    )
            );

    @Override
    public void registerReferenceProviders(@NotNull PsiReferenceRegistrar registrar) {
        PsiReferenceProvider provider = new PsiReferenceProvider() {
            @NotNull
            @Override
            public PsiReference[] getReferencesByElement(@NotNull PsiElement element, @NotNull ProcessingContext context) {
                return new PsiReference[] { new OdooModelReference(element) };
            }
        };
        registrar.registerReferenceProvider(NAME_PATTERN, provider);
        registrar.registerReferenceProvider(INHERIT_PATTERN, provider);
        registrar.registerReferenceProvider(INHERIT_DICT_PATTERN, provider);
        registrar.registerReferenceProvider(INHERITS_PATTERN, provider);
    }
}

File: src/main/kotlin/me/oddlyoko/odoo/models/reference/OdooModelReferenceContributor.kt
package me.oddlyoko.odoo.models.reference

import com.intellij.patterns.PlatformPatterns
import com.intellij.psi.*
import com.intellij.util.ProcessingContext
import com.jetbrains.python.PyTokenTypes
import com.jetbrains.python.psi.*

class OdooModelReferenceContributor: PsiReferenceContributor() {

    override fun registerReferenceProviders(registrar: PsiReferenceRegistrar) {
        val provider = object: PsiReferenceProvider() {
            override fun getReferencesByElement(element: PsiElement, context: ProcessingContext): Array<PsiReference> =
                arrayOf(OdooModelReference(element))
        }
        registrar.registerReferenceProvider(NAME_PATTERN, provider)
        registrar.registerReferenceProvider(INHERIT_PATTERN, provider)
        registrar.registerReferenceProvider(INHERIT_DICT_PATTERN, provider)
        registrar.registerReferenceProvider(INHERITS_PATTERN, provider)
    }

    companion object {
        private val NAME_PATTERN = PlatformPatterns.psiElement(PyStringLiteralExpression::class.java).afterSiblingSkipping(
            PlatformPatterns.psiElement().withElementType(PyTokenTypes.EQ),
            PlatformPatterns.psiElement(PyTargetExpression::class.java).withName("_name")
        )
        private val INHERIT_PATTERN = PlatformPatterns.psiElement(PyStringLiteralExpression::class.java).afterSiblingSkipping(
            PlatformPatterns.psiElement().withElementType(PyTokenTypes.EQ),
            PlatformPatterns.psiElement(PyTargetExpression::class.java).withName("_inherit")
        )
        private val INHERIT_DICT_PATTERN = PlatformPatterns.psiElement(PyStringLiteralExpression::class.java).withParent(
            PlatformPatterns.psiElement(PyListLiteralExpression::class.java).afterSiblingSkipping(
                PlatformPatterns.psiElement().withElementType(PyTokenTypes.EQ),
                PlatformPatterns.psiElement(PyTargetExpression::class.java).withName("_inherit")
            )
        )
        private val INHERITS_PATTERN = PlatformPatterns.psiElement(PyStringLiteralExpression::class.java)
            .withAncestor(2, PlatformPatterns.or(
                PlatformPatterns.psiElement(PyDictLiteralExpression::class.java).afterSiblingSkipping(
                    PlatformPatterns.psiElement().withElementType(PyTokenTypes.EQ),
                    PlatformPatterns.psiElement(PyTargetExpression::class.java).withName("_inherits")
                ),
                PlatformPatterns.psiElement(PySetLiteralExpression::class.java).afterSiblingSkipping(
                    PlatformPatterns.psiElement().withElementType(PyTokenTypes.EQ),
                    PlatformPatterns.psiElement(PyTargetExpression::class.java).withName("_inherits")
                )
            ))
    }
}
